exporter/dynatraceexporter/metrics_exporter.go

```diff
@@ -94,37 +143,225 @@ func (e *exporter) serializeMetrics(md pdata.Metrics) ([]string, int) {
 			metrics := libraryMetric.Metrics()
 			for k := 0; k < metrics.Len(); k++ {
 				metric := metrics.At(k)
-				name, normalizationError := normalizeMetricName(e.cfg.Prefix, metric.Name())
-				if normalizationError != nil {
-					dropped++
-					e.logger.Error(fmt.Sprintf("Failed to normalize metric name: %s", metric.Name()))
-					continue
-				}
 
-				var l []string
+				var metricLines []string
 				switch metric.DataType() {
 				case pdata.MetricDataTypeNone:
 					continue
 				case pdata.MetricDataTypeIntGauge:
-					l = serialization.SerializeIntDataPoints(name, metric.IntGauge().DataPoints(), e.cfg.Tags)
+					for x := 0; x < metric.IntGauge().DataPoints().Len(); x++ {
+						dp := metric.IntGauge().DataPoints().At(x)
+						dims := labelsToDimensions(dp.LabelsMap())
+
+						line, err := serializeMetric(
+							metric.Name(),
+							dtMetric.WithPrefix(e.cfg.Prefix),
+							dtMetric.WithDimensions(
+								dimensions.MergeLists(
+									e.defaultDimensions,
+									dimensions.NewNormalizedDimensionList(dims...),
+									e.staticDimensions,
+								),
+							),
+							dtMetric.WithTimestamp(dp.Timestamp().AsTime()),
+							dtMetric.WithIntGaugeValue(dp.Value()),
+						)
+
+						if err != nil {
+							e.logger.Sugar().Warnf("failed to serialize int gauge %s - %s", metric.Name(), err.Error())
+							continue
+						}
+
+						metricLines = append(metricLines, line)
+					}
 				case pdata.MetricDataTypeDoubleGauge:
-					l = serialization.SerializeDoubleDataPoints(name, metric.DoubleGauge().DataPoints(), e.cfg.Tags)
+					for x := 0; x < metric.DoubleGauge().DataPoints().Len(); x++ {
+						dp := metric.DoubleGauge().DataPoints().At(x)
+						dims := labelsToDimensions(dp.LabelsMap())
+
+						line, err := serializeMetric(
+							metric.Name(),
+							dtMetric.WithPrefix(e.cfg.Prefix),
+							dtMetric.WithDimensions(
+								dimensions.MergeLists(
+									e.defaultDimensions,
+									dimensions.NewNormalizedDimensionList(dims...),
+									e.staticDimensions,
+								),
+							),
+							dtMetric.WithTimestamp(dp.Timestamp().AsTime()),
+							dtMetric.WithFloatGaugeValue(dp.Value()),
+						)
+
+						if err != nil {
+							e.logger.Sugar().Warnf("failed to serialize double gauge %s - %s", metric.Name(), err.Error())
+							continue
+						}
+
+						metricLines = append(metricLines, line)
+					}
 				case pdata.MetricDataTypeIntSum:
-					l = serialization.SerializeIntDataPoints(name, metric.IntSum().DataPoints(), e.cfg.Tags)
+					for x := 0; x < metric.IntSum().DataPoints().Len(); x++ {
+						dp := metric.IntSum().DataPoints().At(x)
+						dims := labelsToDimensions(dp.LabelsMap())
+
+						valueOpt := dtMetric.WithIntCounterValueTotal(dp.Value())
```

Comment:
Is this idiomatic Go? Coming from other languages this looks as if for Deltas two value objects are created, although this only creates the function (option) that would create the object upon creation, right? I assume creating anonymous functions is rather cheap, so I guess this is fine.

Reply:
There is no ternary, which is what I wanted to do. I think it's worth optimizing since it's a tight loop so I'll add an else clause.